src/write.rs

```diff
@@ -11,25 +11,106 @@ use std::io;
 use std::io::prelude::*;
 use std::mem;
 
-#[cfg(any(
-    feature = "deflate",
-    feature = "deflate-miniz",
-    feature = "deflate-zlib"
-))]
-use flate2::write::DeflateEncoder;
+#[cfg(feature = "deflate")]
+mod deflate {
+    use miniz_oxide::deflate::core::{
+        compress_to_output, CompressorOxide, TDEFLFlush, TDEFLStatus as Status,
+    };
+    use miniz_oxide::deflate::CompressionLevel;
+    use std::io;
+
+    pub struct Encoder<W> {
+        inner: W,
+        compressor: Box<CompressorOxide>,
+    }
+    impl<W: io::Write> Encoder<W> {
+        pub fn new(inner: W, level: CompressionLevel) -> Self {
+            let mut compressor: Box<CompressorOxide> = Box::default();
+            compressor.set_format_and_level(miniz_oxide::DataFormat::Raw, level as u8);
+            Self { inner, compressor }
+        }
+        pub fn finish(mut self) -> io::Result<W> {
+            io::Write::flush(&mut self)?;
+            Ok(self.inner)
+        }
+    }
+    impl<W: io::Write> io::Write for Encoder<W> {
+        fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
+            let Self { inner, compressor } = self;
+            let total = buf.len();
+            let mut err = None;
+            loop {
+                match compress_to_output(compressor, buf, TDEFLFlush::None, |mut compressed| loop {
+                    match inner.write(compressed) {
+                        Ok(0) => return true,
+                        Ok(used) => compressed = &compressed[used..],
+                        Err(e) => {
+                            err = Some(e);
+                            return false;
+                        }
+                    }
+                }) {
```

Comment:
This isn't done yet, I was just experimenting with the error handling strategies. Afaik, we can't rewind the compressor, so *something* needs to be done with that `compressed` buffer. We're not allowed to return an error since we may have already written to `inner`, so we'll *probably* need to use an internal buffer to save results here. Maybe we reuse the buffer in the `compressor`?